govcd/vm.go

```diff
@@ -852,6 +853,315 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
+// * a list of IPs
+// * whether the function hit timeout (some IP values may be available after timeout)
+// * error
+// guaranteed to return all IP addresses on a certain timeout.
+//
+// This function checks a slice of nicIndexes and reuses all possible API calls. It may return a
+// partial result for IP addresses when the timeout is hit.
+//
+// Getting a DHCP address is complicated because vCD (in UI and in types.NetworkConnectionSection)
+// reports IP addresses when guest tools are present on a VM only. This function also attempts to
+// check if VM NICs are attached to routed network on edge gateway - then there is a chance that
+// built-in DHCP pools are used and active DHCP leases can be found.
+//
+// For this function to work - at least one the following must be true:
+// * VM has guest tools (vCD UI shows IP address). (Takes longer time)
+// * VM DHCP interface is connected to routed Org network and is using Edge Gateway DHCP. (Takes
+// less time, but is more constrained)
+func (vm *VM) WaitForDhcpIpByNicIndexes(nicIndexes []int, maxWaitSeconds int, useDhcpLeaseCheck bool) ([]string, bool, error) {
+	util.Logger.Printf("[TRACE] VM '%s' - attempting to lookup IP addresses for DHCP NICs %v\n",
+		vm.VM.Name, nicIndexes)
+	// validate NIC indexes
+	if len(nicIndexes) == 0 {
+		return []string{}, false, fmt.Errorf("at least one NIC index must be specified")
+	}
+	for index, nicIndex := range nicIndexes {
+		if nicIndex < 0 {
+			return []string{}, false, fmt.Errorf("NIC index %d cannot be negative", index)
+		}
+	}
+
+	// inject NIC indexes into structure []nicDhcpConfig as this allows to save API calls when
+	// querying for multiple NICs
+	// This slice is ordered the same as original slice of nicIndexes
+	nicStates := make(nicDhcpConfigs, len(nicIndexes))
+	for index, nicIndex := range nicIndexes {
+		nicStates[index].vmNicIndex = nicIndex
+	}
+	var err error
+	if useDhcpLeaseCheck { // Edge gateways have to be looked up when DHCP lease checks are enabled
+		// Lookup edge gateways for routed networks and store them
+		nicStates, err = vm.getEdgeGatewaysForRoutedNics(nicStates)
+		if err != nil {
+			return []string{}, false, fmt.Errorf("unable to validate if NICs are attached to edge gateway: %s", err)
+		}
+	}
+
+	// Run a timer to wait for IPs being present until maxWaitSeconds
+	timeoutAfter := time.After(time.Duration(maxWaitSeconds) * time.Second)
+	tick := time.NewTicker(3 * time.Second)
+	for {
+		select {
+		// If timeout occured - return as much as was found
+		case <-timeoutAfter:
+			ipSlice := getIpsFromNicDhcpConfigs(nicStates)
+			util.Logger.Printf("[DEBUG] VM '%s' NICs with indexes %v did not all report IP "+
+				"addresses after %d seconds. Indexes: %v ,IPs: '%s'\n", vm.VM.Name, nicIndexes,
+				maxWaitSeconds, nicIndexes, strings.Join(ipSlice, ", "))
+			return ipSlice, true, nil
+		case <-tick.C:
+			// Step 1 check if VMware tools reported IPs to UI. Also populate MAC addresses into
```

Comment:
"to **UI**"? Not sure I get how the UI is involved in this call :)

Reply:
UI shows the IPs near NICs when VMware tools report them (as opposed to Step 2 when it is being checked in DHCP leases of edge gateway)

Reply:
I'm confused, because when I read the comment it sounds as if you are somehow querying the UI itself. You're not doing that, right?.. :)

Reply:
No. UI reports IPs from that plance (VMs NetworkConnectionSection property). I have changed the description.

Reply:
Aaa... Maybe then combination of both (possible addition in bold):
> Step 1 check if VMware tools reported IPs in NetworkConnectionSection (**HTML5 UI reads it to show IPs as well).**

Reply:
Added this HTML 5 UI sentence back.